Tools/GMSInfo/source/GMS.cpp

```diff
@@ -0,0 +1,193 @@
+            return false;
+        }
+
+        int result = inflate(&stream, Z_FINISH);
+        inflateEnd(&stream);
+
+        if ((result != Z_STREAM_END || result != Z_BUF_ERROR) && !stream.avail_out) {
+            return false;
+        }
+
+        return true;
+    }
+}
+
+namespace ReGlacier
+{
+    GMS::GMS(std::string name, LevelContainer* levelContainer, LevelAssets* levelAssets)
+        : m_name(std::move(name))
+        , m_container(levelContainer)
+        , m_assets(levelAssets)
+    {}
+
+    void GMS::PrintInfo() {
+        size_t bufferSize = 0;
+        auto buffer = m_container->Read(m_name, bufferSize);
+
+        if (!buffer) {
+            spdlog::error("GMS::PrintInfo() | Unable to read GMS file {}", m_name);
+            return;
+        }
+
+        // Uncompress
+        auto raw = reinterpret_cast<char*>(buffer.get());
+
+        Legacy::GMS2 gms = { 0 };
+        gms.field_0 = 1;
+        gms.m_raw = (int)raw;
+
+        int v5 = *(unsigned long*)raw;
+
+        gms.field_C = v5;
+        gms.field_8 = *(unsigned long*)(raw + 4);
+        gms.field_14 = (*(unsigned char*)(raw + 8)) != 0;
+
+        int outBuffSize = (v5 + 15) & 0xFFFFFFF0;
+
+        auto outBuffer = std::make_unique<char[]>(outBuffSize);
+        Legacy::GMS_Decompress(&gms, outBuffer.get(), outBuffSize);
+
+        spdlog::info("??");
+        EnumerateEntities(std::move(outBuffer));
+    }
+
+    void GMS::EnumerateEntities(std::unique_ptr<char[]>&& gmsBuffer)
+    {
+        std::unique_ptr<char[]> data = std::move(gmsBuffer);
+        char* buffer = data.get();
+
+        size_t prmSize = 0;
+        size_t bufSize = 0;
+
+        auto prm = m_container->Read(m_assets->PRM, prmSize);
+        auto buf = m_container->Read(m_assets->BUF, bufSize);
+
+        if (!prm)
+        {
+            spdlog::error("GMS::EnumerateEntities() Failed to read PRM file {}", m_assets->PRM);
+            return;
+        }
+
+        if (!buf)
+        {
+            spdlog::error("GMS::EnumerateEntities() Failed to read BUF file {}", m_assets->BUF);
+            return;
+        }
+
+#define G_AT(x) (*(((int*)buffer) + x))
+        int flag_0x10 = G_AT(0x10);
+        int flag_0x11 = G_AT(0x11);
+#undef G_AT
+
+        int totalEntitiesCount = *(int*)&buffer[*(int*)buffer];
+        spdlog::info(" ---------------[ GMS ]--------------- ");
+        spdlog::info("Total entities: {}", totalEntitiesCount);
+        spdlog::info("Flag [0x10]: {:X}", flag_0x10);
+        spdlog::info("Flag [0x11]: {:X}", flag_0x11);
+
+        spdlog::info(" AF table");
+        spdlog::info("--------------");
+        int afEntitiesCount = *(int*)(&buf[flag_0x11]);
+        spdlog::info("Total entities: {}", afEntitiesCount);
+
+        int entityNum = 0;
+        char* afPtr = (char*)(&buf[flag_0x11] + sizeof(int)) + 1;
+        std::string_view afid { afPtr };
+
+        // TODO: Loop
+        spdlog::info("[{:4}] {}", entityNum, afid);
```

Comment:
TODO: Think about how to parse it correctly. **afid** string contains all imported animations by their names. Need to find a way to parse it (strings are not separated by null terminator)